metricbeat/module/windows/service/service_windows.go

```diff
@@ -135,6 +146,34 @@ func OpenService(handle ServiceDatabaseHandle, serviceName string, desiredAccess
 	return serviceHandle, nil
 }
 
+func QueryServiceConfig2(serviceHandle ServiceHandle, infoLevel ServiceConfigInformation) ([]byte, error) {
+	var serviceBufSize uint32
+	var serviceBytesNeeded uint32
+
+	if err := _QueryServiceConfig2(serviceHandle, infoLevel, nil, serviceBufSize, &serviceBytesNeeded); err != nil {
+		if ServiceErrno(err.(syscall.Errno)) != SERVICE_ERROR_INSUFFICIENT_BUFFER {
+			err := CloseServiceHandle(serviceHandle)
+			return nil, err
+		}
+		serviceBufSize += serviceBytesNeeded
+		buffer := make([]byte, serviceBufSize)
+
+		for {
```

Comment:
I think you can make this work with a single for loop. The initial call to `_QueryServiceConfig2` before entering the loop should not be necessary. For example see [`getServiceStates`](https://github.com/martinscholz83/beats/blob/0dba05790e37f9a7fe0b88b1cdbbb7100de75903/metricbeat/module/windows/service/service_windows.go#L181).

Reply:
I do the loop initial to don't run into some errors. From the docs
>To determine the required size, specify NULL for this parameter and 0 for the cbBufSize parameter. The function will fail and GetLastError will return ERROR_INSUFFICIENT_BUFFER. The pcbBytesNeeded parameter will receive the required size.

So that's why i do the first run with `nil` as parameter. Or is this unnecessary?

Reply:
I understand why you have done it this way. But there is some duplication that can be eliminated. Consider doing it like this:

```
func QueryServiceConfig2(serviceHandle ServiceHandle, infoLevel ServiceConfigInformation) ([]byte, error) {
	var buffer []byte

	for {
		var bytesNeeded uint32
		var bufPtr *byte
		if len(buffer) > 0 {
			bufPtr = &buffer[0]
		}

		if err := _QueryServiceConfig2(serviceHandle, infoLevel, bufPtr, uint32(len(buffer)), &bytesNeeded); err != nil {
			if ServiceErrno(err.(syscall.Errno)) == SERVICE_ERROR_INSUFFICIENT_BUFFER {
				// Increase buffer size and retry.
				buffer = make([]byte, len(buffer)+int(bytesNeeded))
				continue
			}
			return nil, err
		}
		
		break
	}

	return buffer, nil
}
```

^ * (Disclaimer: I didn't test this at all. ⚠️ )

So in the first iteration a `SERVICE_ERROR_INSUFFICIENT_BUFFER` will happen and the required buffer will be allocated. In the normal case, the second iteration will receive all the data successfully. However the second iteration could fail because something changed (like a new service was added to the OS) and this single loop will handle that too.